Story/70/alex

alexcouper opened the pull request:
@txels This is _far_ from complete, but wanted to let you see it. Any comments welcome.

This is the beginning of solving #70 and #71 

Still needed at the back end to complete this:
- ~~An integration test to ensure we're getting the right hierarchies created.~~ See #77 
- ~~Writing of the `create_hierarchy_trees` (with a possible rename) to calculate which branches from each level are actually merged into higher ones and vice versa. This might end up needing to be moved to within a provider - or at least it will require extensive access to one.~~ See #82

Front end work:
- Extra design work on the front end to have this information displayed in an appropriate way.
- Addition of an angular route to see a feature and its details
- Altering of the feature list to incude urls to the deploystream version of the feature.

Commit:
Have provider specific ids for issues.

Commit:
Add missing test module.

Commit:
Look up id in only the relevant provider.

Commit:
Add hierarchy workings based on regex.

txels commented on deploystream/apps/feature/lib.py:
```diff
@@ -22,42 +23,43 @@ def get_all_features(providers):
-        if feature:
-            break
+    try:
+        planning_provider = providers['by_name'][feature_provider]
```
This may sound weird to you, but why don't we use a single dict that can have different types of keys, either strings when searching for provider name, or classes when searching for provider type. This is Python after all so we are not subject to single-type keys.

I see no danger in overloading the providers dictionary this way, and the code reads clearer.

alexcouper replied:
True - good idea.

txels commented on deploystream/apps/feature/models.py:
```diff
@@ -60,9 +60,12 @@ class Branch(object):
         ``latest_commit`` - The head commmit, or latest revision in this
                             branch.
+        ``level``         - The level this Branch falls in the hierarchy for
+                            the feature.
```
clarify... is this a numeric depth level? Starts at 0?

txels commented on deploystream/apps/feature/views.py:
```diff
@@ -29,9 +30,13 @@ def list_features(providers):
 
 
-@app.route('/features/<feature_id>', methods=['GET'])
+@app.route('/features/<provider_feature_id>', methods=['GET'])
```
Or `<provider>/<feature_id>` ... provider here refers more to origin (the planning provider where we learnt about this feature). 

But as we may get feature info from various providers of other types, I see this naming can be a source of confusion/lack of clarity in the future, once we eventually retrieve feature info specific to a provider (e.g. Jenkins job data for a feature).

alexcouper replied:
Are you saying that the variable name would be a source of confusion to developers, or the url structure to end users?

I'll change to a `/features/<source>/<feature_id>` url structure.

txels replied:
Was thinking developers and variable names yes.

txels commented on deploystream/lib/hierarchy.py:
```diff
@@ -0,0 +1,52 @@
+    """
+    subs = []
+    for index, regex in enumerate(hierarchical_regexes):
+        subs.append("(?P<level_{0}>{1})".format(index, regex))
```
interesting approach

alexcouper replied:
interesting is an ambiguous word - but I'll assume that you're ok with it as is for now...

txels replied:
it was good in this case, something I wouldn't have thought of

txels commented on deploystream/lib/hierarchy.py:
```diff
@@ -0,0 +1,52 @@
+        A list of regexes assumed to be in descending order of branch status.
+
+    :returns:
+        The positional index that the branch should be found it. Or None if it
```
`found in`?

txels commented on deploystream/providers/github/__init__.py:
```diff
@@ -85,3 +86,21 @@ def get_oauth_data(self):
+        branch_list = []
+
+        for repo in self.repositories:
+            for branch in repo.iter_branches():
```
Interesting not to need git clones for this.
I've been thinking of the difficulties of eventually combining git/github providers. Won't be easy to come up with a cleanly layered solution.

txels replied:
(or even just matching Jira/Sprintly projects to actual collections of repos in e.g. Github).

alexcouper replied:
As long as we treat github repos as an independent source from the local git we should be fine. (Where independence means they have their own hierarchy trees etc)

txels commented on deploystream/apps/feature/views.py:
```diff
@@ -29,9 +30,13 @@ def list_features(providers):
-def view_feature(feature_id, providers):
-    feature = get_feature_info(feature_id, providers)
+def view_feature(provider_feature_id, providers):
+    provider_id, feature_id = provider_feature_id.split('...')
```
Strange choice for a separator. Would single `'.'` not work? What about `:` or `;`?

alexcouper replied:
I was obviously being inspired by git at the time - this will change as soon as I make the previous change.

txels commented on deploystream/providers/__init__.py:
```diff
@@ -33,6 +33,7 @@ def get_providers(configs, session):
         interface
     """
     providers = defaultdict(list)
+    providers['by_name'] = {}
```
just stick'em by name, a flat dict.

txels commented:
Looking good. If you wish we can split the work in smaller tasks, as e.g. just the front-end feature detail view seems like quite some work. I would like this to look much nicer than storyboard, so it would be nice to try to sketch out beforehand what we want to display and how (as opposed to getting a bunch of stuff in a template and try to make it look decent as an afterthought).

Commit:
Use a sensible URL and a flat dict.

alexcouper commented:
I think we should absolutely split the work into smaller tasks. I've updated the list of outstanding things in this PR and split them into front end / back end. By all means take the front end ones.
This is the point where github issues begin to show the failings - as it would be ideal to create sub-issues for each of them. Not sure how we want to manage them really.

One way would be to leave this PR open and if you're happy with it simply comment that you are and we then create PRs into this branch for the remaining work and we can therefore see when we're finished and be reminded of the outstanding subtasks...

txels commented:
Good plan.

alexcouper commented on deploystream/providers/github/__init__.py:
```diff
@@ -85,3 +86,21 @@ def get_oauth_data(self):
+
+        for repo in self.repositories:
+            for branch in repo.iter_branches():
+                level = hierarchy.match_with_levels(
```
What do you think about this? It goes against the philosophy that I've been adopting to keep providers as dumb as possible.

An alternative would be that the provider defines a `def iter_branches(self)` method - returning repo name, branch name and head commit - and in https://github.com/pretenders/deploystream/pull/73/files#L0R54 we loop through each performing the relevant `hierarchy.match_with_levels` function.

The second approach would make test writing a lot easier too.

Which way do you prefer?

alexcouper replied:
I think I'm going to take the view that each source code provider is responsible for it's own level producing - and that they may have different ways of calculating that - so therefore leaving the call to `match_with_levels` inside github's provider.

txels replied:
OK I haven't given this enough thought TBH.

Commit:
Add test of get_repo_branches_involved.

Commit:
Add integration test.

Commit:
Change test to import uname and password for testing purposes.

Commit:
Update test.txt requirements.

txels committed:
Merge pull request #77 from pretenders/dev/70/alex

Dev/70/alex

Commit:
Merge branch 'develop' of git+ssh://github.com/pretenders/deploystream into story/70/alex
